content/tutorials/build-apps/build-a-desktop-wallet-in-javascript.md

```diff
@@ -0,0 +1,1433 @@
+
+We also added a new constant containing the directory name where we are going to store our encrypted seed.
+
+3. In `index.js` replace the existing `main` function with the following one:
+
+```javascript
+const main = async () => {
+  const appWindow = createWindow()
+
+  // Create Wallet directory in case it does not exist yet
+  if (!fs.existsSync(WALLET_DIR)) {
+    fs.mkdirSync(path.join(__dirname, WALLET_DIR));
+  }
+
+  let seed = null;
+
+  ipcMain.on('seed-entered', async (event, providedSeed) => {
+    seed = providedSeed
+    appWindow.webContents.send('open-password-dialog')
+  })
+
+  ipcMain.on('password-entered', async (event, password) => {
+    if (!fs.existsSync(path.join(__dirname, WALLET_DIR , 'seed.txt'))) {
+      saveSaltedSeed('../' + WALLET_DIR, seed, password)
+    } else {
+      seed = loadSaltedSeed('../' + WALLET_DIR, password)
+    }
+
+    const wallet = xrpl.Wallet.fromSeed(seed)
+
+    const client = new xrpl.Client(TESTNET_URL)
+
+    await client.connect()
+
+    await subscribe(client, wallet, appWindow)
+
+    await initialize(client, wallet, appWindow)
+  })
+
+  // We have to wait for the application frontend to be ready, otherwise
+  // we might run into a race condition and the open-dialog events
+  // get triggered before the callbacks are attached
+  appWindow.once('ready-to-show', () => {
+    // If there is no seed present yet, ask for it, otherwise query for the password
+    // for the seed that has been saved
+    if (!fs.existsSync(path.join(__dirname, WALLET_DIR, 'seed.txt'))) {
+      appWindow.webContents.send('open-seed-dialog')
+    } else {
+      appWindow.webContents.send('open-password-dialog')
+    }
+  })
+}
+```
+
+The subscription to the `address-entered` frontend is gone, as we are now going to use a full-fledged wallet, for which we are going to prompt the user or a seed and a password to encrypt it with. If there already is a seed, the user will only be asked for his password which is in turn used to decrypt the seed.
+
+3. Then modify the `view/preload.js` file (Note that the `onEnterAccountAddress` function is no longer needed):
+
+```javascript
+contextBridge.exposeInMainWorld('electronAPI', {
+  // Step 5 code additions - start
+  onOpenSeedDialog: (callback) => {
+    ipcRenderer.on('open-seed-dialog', callback)
+  },
+  onEnterSeed: (seed) => {
+    ipcRenderer.send('seed-entered', seed)
+  },
+  onOpenPasswordDialog: (callback) => {
+    ipcRenderer.on('open-password-dialog', callback)
+  },
+  onEnterPassword: (password) => {
+    ipcRenderer.send('password-entered', password)
+  },
+  // Step 5 code additions - end
+
+  onUpdateLedgerData: (callback) => {
+    ipcRenderer.on('update-ledger-data', callback)
+  },
```

Comment:
Closing brackets make it easier to copy + paste over the whole `contextBridge` :)
```suggestion
  },
})
```

Reply:
Nevermind, I'm wrong